File: javascript/94.binary-tree-inorder-traversal.js
/*
 * @lc app=leetcode id=94 lang=javascript
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
  const result = [];
  if (root) {
    if (root.left) {
      result.push(...inorderTraversal(root.left));
    }
    result.push(root.val);
    if (root.right) {
      result.push(...inorderTraversal(root.right));
    }
  }
  return result;
};
// @lc code=end

File: rust/94.binary-tree-inorder-traversal.rs
/*
 * @lc app=leetcode id=94 lang=rust
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = vec![];
        if let Some(node) = root {
            let nb = node.borrow();
            let mut left_nodes = Self::inorder_traversal(nb.left.clone());
            let mut right_nodes = Self::inorder_traversal(nb.right.clone());
            result.append(&mut left_nodes);
            result.push(nb.val);
            result.append(&mut right_nodes);
        }
        return result;
    }
}
// @lc code=end
